Index: webkit-530485e91bc89e8ecb54c07b55e466c9e0c84721/Source/WebKit/WebProcess/WebPage/wpe/AcceleratedSurfaceWPE.cpp
===================================================================
--- webkit-530485e91bc89e8ecb54c07b55e466c9e0c84721.orig/Source/WebKit/WebProcess/WebPage/wpe/AcceleratedSurfaceWPE.cpp
+++ webkit-530485e91bc89e8ecb54c07b55e466c9e0c84721/Source/WebKit/WebProcess/WebPage/wpe/AcceleratedSurfaceWPE.cpp
@@ -75,7 +75,7 @@ void AcceleratedSurfaceWPE::finalize()
 uint64_t AcceleratedSurfaceWPE::window() const
 {
     ASSERT(m_backend);
-    return reinterpret_cast<uint64_t>(wpe_renderer_backend_egl_target_get_native_window(m_backend));
+    return static_cast<uint64_t>(wpe_renderer_backend_egl_target_get_native_window(m_backend));
 }
 
 uint64_t AcceleratedSurfaceWPE::surfaceID() const
