Index: chromium-65b3572/build/linux/unbundle/icu.gn
===================================================================
--- chromium-65b3572.orig/build/linux/unbundle/icu.gn
+++ chromium-65b3572/build/linux/unbundle/icu.gn
@@ -17,6 +17,12 @@ config("icu_config") {
     "USING_SYSTEM_ICU=1",
     "ICU_UTIL_DATA_IMPL=ICU_UTIL_DATA_STATIC",
   ]
+
+  if (is_win) {
+    defines += [ "UCHAR_TYPE=wchar_t" ]
+  } else {
+    defines += [ "UCHAR_TYPE=uint16_t" ]
+  }
 }
 
 pkg_config("system_icui18n") {
